test/performance/odp_crc.c

```diff
@@ -0,0 +1,238 @@
+/* Copyright (c) 2021, Nokia
+ *
+ * All rights reserved.
+ *
+ * SPDX-License-Identifier:     BSD-3-Clause
+ */
+
+#include <stdio.h>
+#include <string.h>
+#include <stdint.h>
+#include <inttypes.h>
+#include <getopt.h>
+
+#include <odp_api.h>
+#include <odp/helper/odph_api.h>
+
+/* Command line options */
+typedef struct {
+	uint32_t time;
+} options_t;
+
+static options_t options;
+static const options_t options_def = {
+	.time = 1000,
+};
+
+static void print_usage(void)
+{
+	printf("\n"
+	       "CRC performance test\n"
+	       "\n"
+	       "Usage: odp_crc_perf [options]\n"
+	       "\n"
+	       "  -t, --time    Approximate time (ms) for each measurement (default %u)\n"
+	       "  -h, --help    This help\n"
+	       "\n",
+	       options_def.time);
+}
+
+static int parse_options(int argc, char *argv[])
+{
+	int opt;
+	int long_index;
+	int ret = 0;
+
+	static const struct option longopts[] = {
+		{ "time", required_argument, NULL, 't' },
+		{ "help", no_argument, NULL, 'h' },
+		{ NULL, 0, NULL, 0 }
+	};
+
+	static const char *shortopts = "+t:h";
+
+	options = options_def;
+
+	while (1) {
+		opt = getopt_long(argc, argv, shortopts, longopts, &long_index);
+
+		if (opt == -1)
+			break;
+
+		switch (opt) {
+		case 't':
+			options.time = atol(optarg);
+			break;
+		case 'h':
+			/* fall through */
+		default:
+			print_usage();
+			ret = -1;
+			break;
+		}
+	}
+
+	if (options.time < 1) {
+		ODPH_ERR("Invalid time: %" PRIu32 "\n", options.time);
+		return -1;
+	}
+
+	return ret;
+}
+
+static uint8_t *buf, *data;
+static const int bufsize = 32 * 1024;
+
+static void prepare_data(void)
+{
+	uint64_t seed = 1;
+	const unsigned long page = 4096;
+
+	/* One extra page for alignment, another one for offset. */
+	buf = (uint8_t *)malloc(bufsize + page * 2);
+	data = (uint8_t *)(((uintptr_t)buf + (page - 1)) & ~(page - 1));
+
+	if (odp_random_test_data(data, bufsize, &seed) != bufsize) {
+		ODPH_ERR("odp_random_test_data() failed.\n");
+		exit(EXIT_FAILURE);
+	}
+}
+
+static void free_data(void)
+{
+	free(buf);
+}
+
+typedef uint32_t (*test_fn_t)(const void *data, uint32_t data_len,
+			      uint32_t init_val);
+
+static uint64_t measure(test_fn_t fn, uint32_t size, uint32_t offset,
+			uint32_t num)
+{
+	void *p = data + offset;
+	volatile uint32_t crc = 1;
```

Comment:
Volatile cannot be used here, since it degrades performance. It tells compiler that something outside of this function may also read/write the variable.

Instead of volatile, output the crc value in the end of the function and e.g. print it out with measurement results. This way compiler cannot optimize crc calculation away, but can perform other optimizations. When I removed volatile, result of the 32B crc32c test case improved 25%.

static uint64_t measure(test_fn_t fn, uint32_t size, uint32_t offset, uint32_t num, uint32_t *crc_out)

Reply:
Fixed in v2 (without the print).